src/chapter_1.md

```diff
@@ -1 +1,409 @@
+
+```rust,ignore
+let initial_source = Box::new(ThePluginSource {
+    random_byte: (self.metrics.expect("Can't read byte_metric")).a_metric,
+});
+
+alumet.add_source(initial_source, TriggerSpec::at_interval(self.config.poll_interval));
+```
+
+In the code above, first, we create a source from ThePluginSource structure, and then we add it to the alumet sources, adding the poll_interval
+value previously defined in the config.
+
+Currently, you should have an error about your initial source, it's because the trait bound
+**ThePluginSource: alumet::pipeline::Source** is not satisfied. Go to implement source to satisfy it.
+
+### Implement Source
+
+In this part, we will implement the Source trait for our ThePluginSource structure.
+
+```rust,ignore
+impl Source for ThePluginSource {
+    fn poll(&mut self, measurements: &mut MeasurementAccumulator, timestamp: Timestamp) -> Result<(), PollError> {
+        ...
+        Ok(())
+    }
+}
+```
+
+This function is called by Alumet each time a measure is needed, so it's in this function that we need to retrieve the value.
+For this example, let's read data from the **/dev/urandom** file.
+Here is the code:
+
+```rust,ignore
+let mut rng = File::open("/dev/urandom")?; // Open the "/dev/urandom" file to obtain random data
+
+let mut buffer = [0u8; 8]; // Create a mutable buffer of type [u8; 8] (an array of 8 unsigned 8-bit integer)
+rng.read_exact(&mut buffer)?; // Read enough byte from the file and store the value in the buffer
+let value: u64 = u64::from_le_bytes(buffer);
+```
+
+> N.b. This will only work on UNIX like OS which does have a file at "/dev/urandom"
+
+We are now able to get the value. The next step is to send this value to ALUMet.
+In order to push data to alumet, we first need to create a measurement point and then push it to the MeasurementAccumulator.
+I also add as an example an attribute the same as value but divided by 2 :
+
+```rust,ignore
+let my_meas_pt = MeasurementPoint::new(
+    timestamp,
+    self.random_byte,
+    Resource::LocalMachine,
+    ResourceConsumer::LocalMachine,
+    value,
+).with_attr("divided", value.div_euclid(2));
+measurements.push(my_meas_pt);
+```
+
+### Add to the app-agent
+
+To add to the app-agent and so use the plugin, several steps are needed.
+
+#### Add the plugin to the cargo.toml
+
+It's very easy, just add your plugin library with a matching version in the toml file:
+
+```toml
+...
+[dependencies]
+...
+my-plugin = {version= "0.1.0", path = "../my-plugin"}
+```
+
+#### Import in the app-agent
+
+In the app-agent main file, import using use:
+
+```rust,ignore
+use my_plugin::ThePlugin;
+```
+
+And then add this newly imported plugin to the statics_plugins macro:
+
+```rust,ignore
+let plugins = static_plugins![ThePlugin, CsvPlugin, SocketControlPlugin];
+```
+
+In this example, we have 3 plugins used by the app-agent:
+
+- ThePlugin
+- CsvPlugin
+- SocketControlPlugin
+
+You can now build ALUMet.
+
+## Final code
```

Comment:
I don't think it's a good idea to have everything in the same file

Reply:
And also I don't know if it's really useful to give the full final code

Reply:
only one example, if i start spliting the code, the user will be lost

Reply:
He won't be lost, but he will push a file with 2000 lines and that's horrible. Every plugins we've written since the beginning are split in multiple files

Reply:
Also, can't you use https://rust-lang.github.io/mdBook/format/mdbook.html#including-portions-of-a-file and https://rust-lang.github.io/mdBook/format/mdbook.html#including-files? This will allow you to write the code in lib.rs and input.rs and then embed some parts inside the doc

Reply:
"you will often wrap the command with ``` to display the file contents without interpreting them"

Reply:
Here, there is 100 lines of code, explained before, split is not necessary. If there is a PR of 2000 lines, just ask to split

Reply:
Here there is only 2 major trait implementations, 4 little structures, there is nothing to split

Reply:
summary of the discussion we've had during the coffee break:
- keep it in one file
- at the end of the tutorial, explain that it's usually best to use another file for the source because it grows in size, and show an example of code of how to do that (using colored diffs would be awesome to highlight the changes!)

Reply:
IIRC this is what the rustbook does :)